File: quicksort.py
def quicksort(lst):
    "Quicksort over a list-like sequence"
    if len(lst) == 0:
        return lst
    pivot = lst[0]
    pivots = [x for x in lst if x == pivot]
    print pivots
    small = quicksort([x for x in lst if x < pivot])
    print small
    large = quicksort([x for x in lst if x > pivot])
    print large
    return small + pivots + large

print quicksort([])
print quicksort([3,2,1])
print quicksort([5,1,0,2,2,4,-1,10])

File: quicksort.rb

def quicksort(arr) 
  return arr if arr.empty?
  pivot = arr.first
  pivots = arr.select {|i| i == pivot }
  small = quicksort arr.select {|i| i < pivot }
  large = quicksort arr.select {|i| i > pivot }
  small + pivots + large
end
p quicksort([])
p quicksort([3,2,1])
p quicksort([5,1,0,2,2,4,-1,10])
